contracts/DaiHrd.sol

```diff
@@ -487,7 +490,57 @@ contract ERC777 is RuntimeConstants, Context, IERC777, IERC20 {
 	}
 }
 
-contract DaiHrd is ERC777 {
+contract MakerFunctions {
+	uint constant ONE = 10 ** 27;
+	function rmul(uint x, uint y) internal pure returns (uint z) {
+		z = mul(x, y) / ONE;
+	}
+
+	function rpow(uint x, uint n, uint base) internal pure returns (uint z) {
+		assembly {
+			switch x case 0 {switch n case 0 {z := base} default {z := 0}}
+			default {
+				switch mod(n, 2) case 0 { z := base } default { z := x }
+				let half := div(base, 2)  // for rounding.
+				for { n := div(n, 2) } n { n := div(n,2) } {
+				let xx := mul(x, x)
+				if iszero(eq(div(xx, x), x)) { revert(0,0) }
+				let xxRound := add(xx, half)
+				if lt(xxRound, xx) { revert(0,0) }
+				x := div(xxRound, base)
+				if mod(n,2) {
+					let zx := mul(z, x)
+					if and(iszero(iszero(x)), iszero(eq(div(zx, x), z))) { revert(0,0) }
+					let zxRound := add(zx, half)
+					if lt(zxRound, zx) { revert(0,0) }
+					z := div(zxRound, base)
+				}
+			}
+			}
+		}
+	}
+
+	function mul(uint x, uint y) internal pure returns (uint z) {
+		require(y == 0 || (z = x * y) / y == x);
+	}
+
+	function calculatedChi(Pot pot) internal view returns (uint256 rontodaiPerPot) {
+		rontodaiPerPot = rmul(rpow(pot.dsr(), now - pot.rho(), ONE), pot.chi());
+		return rontodaiPerPot;
+	}
+
+	function updateAndFetchChi(Pot pot) internal returns (uint256 rontodaiPerPot) {
+		if (pot.rho() == now) {
+			rontodaiPerPot = pot.chi();
+		} else {
+			rontodaiPerPot = pot.drip();
+		}
+		return rontodaiPerPot;
+	}
+}
```

Comment:
I'm assuming these are just copy-pasta from Maker?

Reply:
one, rpow, rmul, mul are copies. The logic in calculatedChi is lifted.

But rontodaiPerPot is us, no one else uses that term. I should bring these chi ones back to the main contract, since they stopped looking like Maker code